### PR TITLE
rustc_llvm: Fix flattened CLI args

### PR DESCRIPTION
**Pull request by aeubanks**
Fixes string manipulation errors introduced in #130446.

### CODE REVIEW

**Comment by aeubanks** [comment]
r? augie

**Comment by aeubanks** [comment]
r? durin42

**Comment by workingjubilee** [comment]
Thanks.

I'm... actually not sure how rustc finished bootstrap on the LLVM 20 integration build with this error? It _has_ built since then, right?

**Comment by aeubanks** [comment]
This only affects the command line portion in debug info on windows AFAIK. The only reason we found this was because we had tests for deterministic builds on windows, and the bad strings were sometimes garbage.

**Comment by workingjubilee** [comment]
```
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xbb2be6)[0x7fb88cd3fbe6]
/lib/x86_64-linux-gnu/libc.so.6(+0x3c050)[0x7fb88bfe2050]
/lib/x86_64-linux-gnu/libc.so.6(+0x15b33c)[0x7fb88c10133c]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(LLVMRustCreateTargetMachine+0x494)[0x7fb88d224a54]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0x101decb)[0x7fb88d1aaecb]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0x101ca7e)[0x7fb88d1a9a7e]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xf52efd)[0x7fb88d0dfefd]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0x1024ded)[0x7fb88d1b1ded]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0x102428f)[0x7fb88d1b128f]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xf618d1)[0x7fb88d0ee8d1]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(_RNvXs5_Cs1wnS3Bmr5sj_18rustc_codegen_llvmNtB5_18LlvmCodegenBackendNtNtNtCsly6bReQAbem_17rustc_codegen_ssa6traits7backend14CodegenBackend13codegen_crate+0xca)[0x7fb88d0ce5ba]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xee55d7)[0x7fb88d0725d7]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xe5c2e0)[0x7fb88cfe92e0]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(_RNvMs3_NtCslONYkxYabyj_15rustc_interface7queriesNtB5_6Linker24codegen_and_build_linker+0x22)[0x7fb88d094de2]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xbbb2b5)[0x7fb88cd482b5]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xbb37a5)[0x7fb88cd407a5]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xb81293)[0x7fb88cd0e293]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xbc4a7b)[0x7fb88cd51a7b]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xbbcc82)[0x7fb88cd49c82]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0xb7cfd7)[0x7fb88cd09fd7]
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-ad72a12415e858c4.so(+0x6c6f7db)[0x7fb892dfc7db]
/lib/x86_64-linux-gnu/libc.so.6(+0x89134)[0x7fb88c02f134]
/lib/x86_64-linux-gnu/libc.so.6(__clone+0x40)[0x7fb88c0aea40]
note: we would appreciate a report at https://github.com/rust-lang/rust
help: you can increase rustc's stack size by setting RUST_MIN_STACK=16777216
```

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp [hunk 1]
```diff
@@ -490,13 +490,13 @@ extern "C" LLVMTargetMachineRef LLVMRustCreateTargetMachine(
...
-    auto ArgsCppStr =
-        std::string(ArgsCstrBuff + buffer_offset, ArgsCstrBuffLen - 1);
+    auto ArgsCppStr = std::string(ArgsCstrBuff + buffer_offset,
+                                  ArgsCstrBuffLen - buffer_offset);
     auto i = 0;
     while (i != std::string::npos) {
       i = ArgsCppStr.find('\0', i + 1);
       if (i != std::string::npos)
-        ArgsCppStr.replace(i, i + 1, " ");
+        ArgsCppStr.replace(i, 1, " ");
```
Comment (workingjubilee) [inline compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:499]: So, re-reading the reference about `std::string::npos`, I think I may have misunderstood something about it.

Why is it correct at all as part of a conditional expression like a `while` or `if`? It only takes on its magic "until the end of the string" meaning when used as an argument to `std::string`'s functions, doesn't it? Which the control-flow expressions are not invocations of.

Comment (nico) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:499]: https://cplusplus.com/reference/string/string/find/

> Return Value
> The position of the first character of the first match.
> If no matches were found, the function returns [string::npos](https://cplusplus.com/string::npos).

`npos` isn't magic, it's in practice just a way to say `-1`. find() returns npos if something isn't found, and at that point we don't take this if / exit the loop.

Comment (workingjubilee) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:499]: Ah, okay!

**Pull request merged**